presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometrySerde.java

```diff
@@ -205,51 +203,94 @@ public static OGCGeometry createFromEsriGeometry(Geometry geometry, boolean mult
     public static Envelope deserializeEnvelope(Slice shape)
     {
-        if (shape == null) {
+        requireNonNull(shape, "shape is null");
+        BasicSliceInput input = shape.getInput();
+
+        if (input.available() == 0) {
             return null;
         }
-        BasicSliceInput input = shape.getInput();
 
-        Envelope overallEnvelope = null;
-        if (input.available() > 0) {
-            byte code = input.readByte();
-            boolean isGeometryCollection = (code == GeometryType.GEOMETRY_COLLECTION.code());
-            while (input.available() > 0) {
-                int length = isGeometryCollection ? input.readInt() : input.available();
-                ByteBuffer buffer = input.readSlice(length).toByteBuffer().order(LITTLE_ENDIAN);
-                int type = buffer.getInt();
-                Envelope envelope = null;
-                if (type == POINT_TYPE) {    // point
-                    double x = buffer.getDouble();
-                    double y = buffer.getDouble();
-                    if (!GeometryUtils.isEsriNaN(x)) {
-                        verify(!GeometryUtils.isEsriNaN(y));
-                        envelope = new Envelope(x, y, x, y);
-                    }
-                }
-                else {
-                    double xMin = buffer.getDouble();
-                    double yMin = buffer.getDouble();
-                    double xMax = buffer.getDouble();
-                    double yMax = buffer.getDouble();
-                    if (!GeometryUtils.isEsriNaN(xMin)) {
-                        verify(!GeometryUtils.isEsriNaN(xMax));
-                        verify(!GeometryUtils.isEsriNaN(yMin));
-                        verify(!GeometryUtils.isEsriNaN(yMax));
-                        envelope = new Envelope(xMin, yMin, xMax, yMax);
-                    }
-                }
-                if (envelope != null) {
-                    if (overallEnvelope == null) {
-                        overallEnvelope = envelope;
-                    }
-                    else {
-                        overallEnvelope.merge(envelope);
-                    }
-                }
-            }
+        int length = input.available() - 1;
+        GeometryType type = GeometryType.getForCode(input.readByte());
+        return getEnvelope(input, type, length);
+    }
+
+    private static Envelope getEnvelope(BasicSliceInput input, GeometryType type, int length)
+    {
+        switch (type) {
+            case POINT:
+                return getPointEnvelope(input);
+            case MULTI_POINT:
+            case LINE_STRING:
+            case MULTI_LINE_STRING:
+            case POLYGON:
+            case MULTI_POLYGON:
+                return getSimpleGeometryEnvelope(input, length);
+            case GEOMETRY_COLLECTION:
+                return getGeometryCollectionOverallEnvelope(input);
+            default:
+                throw new UnsupportedOperationException("Unexpected type: " + type);
```

Comment:
In other switches `IllegalArgumentException` is thrown, which I think is a better fit to here.